core/lib/zksync_core/src/sync_layer/gossip/storage.rs

```diff
@@ -0,0 +1,351 @@
+        let mut storage = self.storage().await?;
+        let number = storage
+            .blocks_dal()
+            .get_sealed_miniblock_number()
+            .await
+            .context("Failed getting sealed miniblock number")?;
+        Ok(BlockNumber(number.0.into()))
+    }
+}
+
+#[async_trait]
+impl BlockStore for PostgresBlockStorage {
+    async fn head_block(&self, ctx: &ctx::Ctx) -> StorageResult<FinalBlock> {
+        tokio::select! {
+            () = ctx.canceled() => Err(ctx::Canceled.into()),
+            result = self.head_block() => result.map_err(StorageError::Database),
+        }
+    }
+
+    async fn first_block(&self, ctx: &ctx::Ctx) -> StorageResult<FinalBlock> {
+        tokio::select! {
+            () = ctx.canceled() => Err(ctx::Canceled.into()),
+            result = self.first_block() => result.map_err(StorageError::Database),
+        }
+    }
+
+    async fn last_contiguous_block_number(&self, ctx: &ctx::Ctx) -> StorageResult<BlockNumber> {
+        tokio::select! {
+            () = ctx.canceled() => Err(ctx::Canceled.into()),
+            result = self.sealed_miniblock_number() => result.map_err(StorageError::Database),
+        }
+    }
+
+    async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: BlockNumber,
+    ) -> StorageResult<Option<FinalBlock>> {
+        let get_block = async {
+            let number = u32::try_from(number.0).context("block number is too large")?;
+            let mut storage = self.storage().await?;
+            Self::block(&mut storage, MiniblockNumber(number)).await
+        };
+        tokio::select! {
+            () = ctx.canceled() => Err(ctx::Canceled.into()),
+            result = get_block => result.map_err(StorageError::Database),
+        }
+    }
+
+    async fn missing_block_numbers(
+        &self,
+        _ctx: &ctx::Ctx,
+        _range: ops::Range<BlockNumber>,
+    ) -> StorageResult<Vec<BlockNumber>> {
+        Ok(vec![]) // The storage never has missing blocks by construction
+    }
+
+    fn subscribe_to_block_writes(&self) -> watch::Receiver<BlockNumber> {
+        self.block_sender.subscribe()
+    }
+}
+
+#[async_trait]
+impl ContiguousBlockStore for PostgresBlockStorage {
+    async fn schedule_next_block(&self, ctx: &ctx::Ctx, block: &FinalBlock) -> StorageResult<()> {
+        let fetched_block =
+            FetchedBlock::from_gossip_block(block).map_err(StorageError::Database)?;
+        let actions = sync::lock(ctx, &self.cursor).await?.advance(fetched_block);
+        let push_all_actions = async {
+            for actions_chunk in actions {
+                self.actions.push_actions(actions_chunk).await;
+            }
+        };
+        tokio::select! {
+            () = ctx.canceled() => Err(ctx::Canceled.into()),
+            () = push_all_actions => Ok(()),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::{thread_rng, Rng};
+
+    use zksync_concurrency::scope;
+    use zksync_types::L2ChainId;
```

Comment:
I know, that zksync-era isn't very consistent on that, but can we move towards having tests in a separate file?